File: backend/services/carpool_matching.py
"""
Carpool Matching Service
Port of frontend carpoolMatching.ts to Python.
Haversine distance, route/time/preference scoring.
"""

import math
from typing import Optional, Dict, List, Any


def haversine_distance(lat1: float, lng1: float, lat2: float, lng2: float) -> float:
    """Calculate distance between two GPS coordinates in km using Haversine formula."""
    R = 6371  # Earth's radius in km
    d_lat = math.radians(lat2 - lat1)
    d_lng = math.radians(lng2 - lng1)

    a = (
        math.sin(d_lat / 2) ** 2
        + math.cos(math.radians(lat1))
        * math.cos(math.radians(lat2))
        * math.sin(d_lng / 2) ** 2
    )
    c = 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))
    return R * c


def time_to_minutes(time_str: str) -> int:
    """Convert HH:MM string to minutes since midnight."""
    parts = time_str.split(":")
    return int(parts[0]) * 60 + int(parts[1])


def calculate_route_score(
    user_origin_lat: float,
    user_origin_lng: float,
    user_dest_lat: float,
    user_dest_lng: float,
    ride_origin_lat: float,
    ride_origin_lng: float,
    ride_dest_lat: float,
    ride_dest_lng: float,
    flexible_radius: float = 5.0,
) -> float:
    """Score route compatibility (0-100)."""
    origin_dist = haversine_distance(user_origin_lat, user_origin_lng, ride_origin_lat, ride_origin_lng)
    dest_dist = haversine_distance(user_dest_lat, user_dest_lng, ride_dest_lat, ride_dest_lng)

    origin_score = max(0, 100 - (origin_dist / flexible_radius) * 100)
    dest_score = max(0, 100 - (dest_dist / flexible_radius) * 100)

    return (origin_score + dest_score) / 2


def calculate_time_score(
    user_time: str,
    ride_time: str,
    flexible_minutes: int = 30,
) -> float:
    """Score time compatibility (0-100)."""
    user_min = time_to_minutes(user_time)
    ride_min = time_to_minutes(ride_time)
    diff = abs(user_min - ride_min)

    if diff == 0:
        return 100
    if diff > flexible_minutes * 2:
        return 0
    return max(0, 100 - (diff / flexible_minutes) * 50)


def calculate_preference_score(
    user_prefs: Dict[str, Any],
    ride_prefs: Dict[str, Any],
) -> float:
    """Score preference compatibility (0-100)."""
    score = 0

    # Music (20 pts)
    u_music = user_prefs.get("musicPreference", "no-preference")
    r_music = ride_prefs.get("musicPreference", "no-preference")
    if u_music == "no-preference" or r_music == "no-preference":
        score += 20
    elif u_music == r_music:
        score += 20
    else:
        score += 10

    # Conversation (20 pts)
    u_conv = user_prefs.get("conversationLevel", "no-preference")
    r_conv = ride_prefs.get("conversationLevel", "no-preference")
    if u_conv == "no-preference" or r_conv == "no-preference":
        score += 20
    elif u_conv == r_conv:
        score += 20
    else:
        score += 10

    # Pets (15 pts)
    if user_prefs.get("allowsPets") == ride_prefs.get("allowsPets"):
        score += 15
    elif not user_prefs.get("allowsPets") and ride_prefs.get("allowsPets"):
        score += 0
    else:
        score += 7

    # Smoking (25 pts)
    u_smoke = user_prefs.get("allowsSmoking", False)
    r_smoke = ride_prefs.get("allowsSmoking", False)
    if not u_smoke and not r_smoke:
        score += 25
    elif u_smoke == r_smoke:
        score += 25
    elif not u_smoke and r_smoke:
        score += 0
    else:
        score += 15

    # Vehicle type (20 pts)
    u_veh = user_prefs.get("vehicleType", "any")
    r_veh = ride_prefs.get("vehicleType", "any")
    if u_veh == "any" or r_veh == "any":
        score += 20
    elif u_veh == r_veh:
        score += 20
    elif u_veh in ("electric", "hybrid") and r_veh in ("electric", "hybrid"):
        score += 15
    else:
        score += 10

    return score


def calculate_compatibility_score(
    user_origin: tuple,
    user_dest: tuple,
    user_time: str,
    ride_origin: tuple,
    ride_dest: tuple,
    ride_time: str,
    user_prefs: Dict = None,
    ride_prefs: Dict = None,
    driver_rating: Optional[float] = None,
    weights: Dict = None,
) -> Dict:
    """
    Calculate overall match score.
    Returns dict with overall score and component breakdowns.
    """
    if weights is None:
        weights = {
            "route": 0.4,
            "time": 0.3,
            "preferences": 0.2,
            "rating": 0.1,
        }

    route_score = calculate_route_score(
        user_origin[0], user_origin[1],
        user_dest[0], user_dest[1],
        ride_origin[0], ride_origin[1],
        ride_dest[0], ride_dest[1],
    )

    time_score = calculate_time_score(user_time, ride_time)

    pref_score = 80  # Default
    if user_prefs and ride_prefs:
        pref_score = calculate_preference_score(user_prefs, ride_prefs)

    rating_score = (driver_rating / 5 * 100) if driver_rating else 80

    overall = (
        route_score * weights["route"]
        + time_score * weights["time"]
        + pref_score * weights["preferences"]
        + rating_score * weights["rating"]
    )

    return {
        "overall_score": round(overall),
        "route_score": round(route_score, 1),
        "time_score": round(time_score, 1),
        "preference_score": round(pref_score, 1),
        "rating_score": round(rating_score, 1),
    }


def get_match_quality(score: int) -> Dict[str, str]:
    """Get match quality label and description."""
    if score >= 90:
        return {"label": "Excellent Match", "color": "green", "description": "Highly compatible route and preferences"}
    elif score >= 80:
        return {"label": "Great Match", "color": "green", "description": "Very good compatibility"}
    elif score >= 70:
        return {"label": "Good Match", "color": "blue", "description": "Compatible with minor differences"}
    elif score >= 60:
        return {"label": "Fair Match", "color": "yellow", "description": "Some compatibility concerns"}
    else:
        return {"label": "Low Match", "color": "red", "description": "Limited compatibility"}

File: frontend/src/app/utils/carpoolMatching.ts
// Carpool Matching Algorithm Utilities

export interface Location {
  lat: number;
  lng: number;
  address: string;
}

export interface CommutePreferences {
  musicPreference: 'quiet' | 'music' | 'podcast' | 'no-preference';
  conversationLevel: 'chatty' | 'quiet' | 'no-preference';
  allowsPets: boolean;
  allowsSmoking: boolean;
  vehicleType: 'sedan' | 'suv' | 'electric' | 'hybrid' | 'any';
  temperaturePreference?: 'warm' | 'cool' | 'no-preference';
  flexibleTiming: boolean;
  flexibleRadius: number; // km
}

export interface MatchingCriteria {
  routeWeight: number; // 0-1
  timeWeight: number; // 0-1
  preferencesWeight: number; // 0-1
  ratingWeight: number; // 0-1
}

export interface RouteInfo {
  origin: Location;
  destination: Location;
  departureTime: string;
  distance: number;
}

const DEFAULT_MATCHING_CRITERIA: MatchingCriteria = {
  routeWeight: 0.4,
  timeWeight: 0.3,
  preferencesWeight: 0.2,
  ratingWeight: 0.1,
};

/**
 * Calculate distance between two geographic points using Haversine formula
 * @returns distance in kilometers
 */
export function calculateDistance(
  lat1: number,
  lng1: number,
  lat2: number,
  lng2: number
): number {
  const R = 6371; // Earth's radius in km
  const dLat = toRadians(lat2 - lat1);
  const dLng = toRadians(lng2 - lng1);

  const a =
    Math.sin(dLat / 2) * Math.sin(dLat / 2) +
    Math.cos(toRadians(lat1)) *
      Math.cos(toRadians(lat2)) *
      Math.sin(dLng / 2) *
      Math.sin(dLng / 2);

  const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
  return R * c;
}

function toRadians(degrees: number): number {
  return degrees * (Math.PI / 180);
}

/**
 * Calculate route compatibility score (0-100)
 */
export function calculateRouteScore(
  userRoute: RouteInfo,
  rideRoute: RouteInfo,
  flexibleRadius: number = 5
): number {
  // Calculate distance between origins
  const originDistance = calculateDistance(
    userRoute.origin.lat,
    userRoute.origin.lng,
    rideRoute.origin.lat,
    rideRoute.origin.lng
  );

  // Calculate distance between destinations
  const destDistance = calculateDistance(
    userRoute.destination.lat,
    userRoute.destination.lng,
    rideRoute.destination.lat,
    rideRoute.destination.lng
  );

  // Score based on how close both origin and destination are
  const originScore = Math.max(0, 100 - (originDistance / flexibleRadius) * 100);
  const destScore = Math.max(0, 100 - (destDistance / flexibleRadius) * 100);

  // Average of both scores
  return (originScore + destScore) / 2;
}

/**
 * Calculate time compatibility score (0-100)
 */
export function calculateTimeScore(
  userTime: string,
  rideTime: string,
  flexibleMinutes: number = 30
): number {
  const userMinutes = timeToMinutes(userTime);
  const rideMinutes = timeToMinutes(rideTime);
  const timeDiff = Math.abs(userMinutes - rideMinutes);

  if (timeDiff === 0) return 100;
  if (timeDiff > flexibleMinutes * 2) return 0;

  return Math.max(0, 100 - (timeDiff / flexibleMinutes) * 50);
}

function timeToMinutes(time: string): number {
  const [hours, minutes] = time.split(':').map(Number);
  return hours * 60 + minutes;
}

/**
 * Calculate preferences compatibility score (0-100)
 */
export function calculatePreferencesScore(
  userPrefs: CommutePreferences,
  ridePrefs: CommutePreferences
): number {
  let score = 0;
  let criteriaCount = 0;

  // Music preference (20 points)
  if (userPrefs.musicPreference === 'no-preference' || ridePrefs.musicPreference === 'no-preference') {
    score += 20;
  } else if (userPrefs.musicPreference === ridePrefs.musicPreference) {
    score += 20;
  } else if (
    (userPrefs.musicPreference === 'quiet' && ridePrefs.musicPreference !== 'music') ||
    (ridePrefs.musicPreference === 'quiet' && userPrefs.musicPreference !== 'music')
  ) {
    score += 10;
  }
  criteriaCount += 20;

  // Conversation level (20 points)
  if (userPrefs.conversationLevel === 'no-preference' || ridePrefs.conversationLevel === 'no-preference') {
    score += 20;
  } else if (userPrefs.conversationLevel === ridePrefs.conversationLevel) {
    score += 20;
  } else {
    score += 10;
  }
  criteriaCount += 20;

  // Pets (15 points)
  if (userPrefs.allowsPets === ridePrefs.allowsPets) {
    score += 15;
  } else if (!userPrefs.allowsPets && ridePrefs.allowsPets) {
    // Passenger doesn't want pets but driver allows them
    score += 0;
  } else {
    score += 7;
  }
  criteriaCount += 15;

  // Smoking (25 points - most important for safety/health)
  if (!userPrefs.allowsSmoking && !ridePrefs.allowsSmoking) {
    score += 25;
  } else if (userPrefs.allowsSmoking && ridePrefs.allowsSmoking) {
    score += 25;
  } else if (!userPrefs.allowsSmoking && ridePrefs.allowsSmoking) {
    score += 0; // Deal breaker for non-smokers
  } else {
    score += 15;
  }
  criteriaCount += 25;

  // Vehicle type (20 points)
  if (userPrefs.vehicleType === 'any' || ridePrefs.vehicleType === 'any') {
    score += 20;
  } else if (userPrefs.vehicleType === ridePrefs.vehicleType) {
    score += 20;
  } else if (
    (userPrefs.vehicleType === 'electric' || userPrefs.vehicleType === 'hybrid') &&
    (ridePrefs.vehicleType === 'electric' || ridePrefs.vehicleType === 'hybrid')
  ) {
    score += 15;
  } else {
    score += 10;
  }
  criteriaCount += 20;

  return score;
}

/**
 * Calculate overall compatibility score
 */
export function calculateCompatibilityScore(
  userRoute: RouteInfo,
  rideRoute: RouteInfo,
  userPrefs: CommutePreferences,
  ridePrefs: CommutePreferences,
  driverRating?: number,
  criteria: MatchingCriteria = DEFAULT_MATCHING_CRITERIA
): number {
  const routeScore = calculateRouteScore(userRoute, rideRoute, userPrefs.flexibleRadius);
  const timeScore = calculateTimeScore(
    userRoute.departureTime,
    rideRoute.departureTime,
    userPrefs.flexibleTiming ? 30 : 15
  );
  const prefsScore = calculatePreferencesScore(userPrefs, ridePrefs);
  const ratingScore = driverRating ? (driverRating / 5) * 100 : 80; // Default to 80 if no rating

  const overallScore =
    routeScore * criteria.routeWeight +
    timeScore * criteria.timeWeight +
    prefsScore * criteria.preferencesWeight +
    ratingScore * criteria.ratingWeight;

  return Math.round(overallScore);
}

/**
 * Get match quality label
 */
export function getMatchQuality(score: number): {
  label: string;
  color: string;
  description: string;
} {
  if (score >= 90) {
    return {
      label: 'Excellent Match',
      color: 'green',
      description: 'Highly compatible route and preferences',
    };
  } else if (score >= 80) {
    return {
      label: 'Great Match',
      color: 'green',
      description: 'Very good compatibility',
    };
  } else if (score >= 70) {
    return {
      label: 'Good Match',
      color: 'blue',
      description: 'Compatible with minor differences',
    };
  } else if (score >= 60) {
    return {
      label: 'Fair Match',
      color: 'yellow',
      description: 'Some compatibility concerns',
    };
  } else {
    return {
      label: 'Low Match',
      color: 'red',
      description: 'Limited compatibility',
    };
  }
}

/**
 * Filter rides by advanced criteria
 */
export interface AdvancedFilters {
  maxDistance?: number;
  departureWindowStart?: string;
  departureWindowEnd?: string;
  minSeats?: number;
  minRating?: number;
  vehicleTypes?: string[];
  mustHavePreferences?: Partial<CommutePreferences>;
}

export function applyAdvancedFilters<T extends { distance: number; departureTime: string; seatsAvailable: number; driverRating?: number; vehicleType?: string }>(
  rides: T[],
  filters: AdvancedFilters
): T[] {
  return rides.filter((ride) => {
    // Distance filter
    if (filters.maxDistance && ride.distance > filters.maxDistance) {
      return false;
    }

    // Time window filter
    if (filters.departureWindowStart || filters.departureWindowEnd) {
      const rideMinutes = timeToMinutes(ride.departureTime);
      if (filters.departureWindowStart) {
        const startMinutes = timeToMinutes(filters.departureWindowStart);
        if (rideMinutes < startMinutes) return false;
      }
      if (filters.departureWindowEnd) {
        const endMinutes = timeToMinutes(filters.departureWindowEnd);
        if (rideMinutes > endMinutes) return false;
      }
    }

    // Seats filter
    if (filters.minSeats && ride.seatsAvailable < filters.minSeats) {
      return false;
    }

    // Rating filter
    if (filters.minRating && ride.driverRating && ride.driverRating < filters.minRating) {
      return false;
    }

    // Vehicle type filter
    if (filters.vehicleTypes && filters.vehicleTypes.length > 0 && ride.vehicleType) {
      if (!filters.vehicleTypes.includes(ride.vehicleType)) {
        return false;
      }
    }

    return true;
  });
}

/**
 * Generate mock location data (for demo purposes)
 */
export function geocodeAddress(address: string): Location {
  // Mock geocoding - in production, use Google Maps API or similar
  const locations: Record<string, { lat: number; lng: number }> = {
    // Dublin / Ireland
    'Acme Dublin HQ': { lat: 53.3498, lng: -6.2603 },
    'Acme Dublin HQ, Dublin 2': { lat: 53.3498, lng: -6.2603 },
    'Dublin City Centre': { lat: 53.3498, lng: -6.2603 },
    'Dublin 2': { lat: 53.3340, lng: -6.2535 },
    'Griffith Avenue DART Stop, Dublin 9': { lat: 53.3736, lng: -6.2521 },
    'Griffith Avenue': { lat: 53.3736, lng: -6.2521 },
    'Sandyford': { lat: 53.2745, lng: -6.2157 },
    'Salthill Road, Galway': { lat: 53.2635, lng: -9.0850 },
    'Acme Galway Campus, Galway': { lat: 53.2707, lng: -9.0568 },
    'Acme Galway Campus': { lat: 53.2707, lng: -9.0568 },
    'Acme Cork Office, Cork': { lat: 51.8979, lng: -8.4706 },
    'Leopardstown': { lat: 53.2745, lng: -6.2157 },
    'Clonskeagh': { lat: 53.3047, lng: -6.2325 },
    'Ranelagh': { lat: 53.3250, lng: -6.2594 },
    // London / UK
    'London Office': { lat: 51.5074, lng: -0.1278 },
    'Canary Wharf': { lat: 51.5049, lng: -0.0236 },
    // Default (Dublin centre)
    'Home': { lat: 53.3498, lng: -6.2603 },
    'Work': { lat: 53.3340, lng: -6.2535 },
  };

  const coords = locations[address] || { lat: 53.3498, lng: -6.2603 };

  return {
    ...coords,
    address,
  };
}

/**
 * Calculate estimated CO2 savings
 */
export function calculateCO2Savings(distance: number, passengers: number = 1): number {
  // Average car emissions: 0.21 kg CO2 per km
  // Carpooling divides emissions by (passengers + driver)
  const soloEmissions = distance * 0.21;
  const carpoolEmissions = soloEmissions / (passengers + 1);
  const savings = soloEmissions - carpoolEmissions;
  return Math.round(savings * 10) / 10;
}
